File: producer-consumer-rabbitmq/consumer-java/src/main/java/Consumer.java
import com.rabbitmq.client.*;

public class Consumer {
    private final static String QUEUE_NAME = "deportes";

    public static void main(String[] argv) {
        int maxRetries = 10;   // número máximo de intentos
        int waitTime = 5000;   // tiempo entre intentos (ms)

        Connection connection = null;
        Channel channel = null;

        for (int attempt = 1; attempt <= maxRetries; attempt++) {
            try {
                ConnectionFactory factory = new ConnectionFactory();
                factory.setHost("rabbitmq");   // 👈 nombre del servicio docker-compose
                factory.setPort(5672);
                factory.setUsername("user");
                factory.setPassword("password");

                connection = factory.newConnection();
                channel = connection.createChannel();

                channel.queueDeclare(QUEUE_NAME, false, false, false, null);
                System.out.println("[✔] Conectado a RabbitMQ. Esperando mensajes...");

                DeliverCallback deliverCallback = (consumerTag, delivery) -> {
                    String message = new String(delivery.getBody(), "UTF-8");
                    System.out.println("[x] Recibido: '" + message + "'");
                };

                channel.basicConsume(QUEUE_NAME, true, deliverCallback, consumerTag -> {});

                // 🚀 Importante: salimos del bucle si la conexión fue exitosa
                break;

            } catch (Exception e) {
                System.out.println("[!] RabbitMQ no está listo. Intento " + attempt + "/" + maxRetries);
                try {
                    Thread.sleep(waitTime);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                }
            }
        }

        if (connection == null) {
            System.err.println("[✘] No se pudo conectar a RabbitMQ después de varios intentos.");
            System.exit(1);
        }
    }
}

File: producer-consumer-rabbitmq/consumer-py/consumer.py
import pika, time

QUEUE_NAME = "tareas"

def connect_rabbitmq(max_retries=10, wait_time=5):
    for attempt in range(max_retries):
        try:
            connection = pika.BlockingConnection(
                pika.ConnectionParameters(
                    host="rabbitmq",
                    port=5672,
                    credentials=pika.PlainCredentials("user", "password")
                )
            )
            print("[✔] Conectado a RabbitMQ (Python Consumer)")
            return connection
        except pika.exceptions.AMQPConnectionError:
            print(f"[!] RabbitMQ no está listo. Intento {attempt+1}/{max_retries}")
            time.sleep(wait_time)
    raise Exception("No se pudo conectar a RabbitMQ después de varios intentos")

def callback(ch, method, properties, body):
    print(f"[x] Recibido: {body.decode()}")

connection = connect_rabbitmq()
channel = connection.channel()
channel.queue_declare(queue=QUEUE_NAME)
channel.basic_consume(queue=QUEUE_NAME, on_message_callback=callback, auto_ack=True)

print("[✔] Esperando mensajes...")
channel.start_consuming()
